src/ZoneServer/Database/ZoneDb.cs

```diff
@@ -1205,5 +1207,113 @@ private void LoadQuests(Character character)
 				}
 			}
 		}
+			
+
+		/// <summary>
+		/// Saves the characters's collections to the database.
+		/// Note that collections are account-level
+		/// </summary>
+		/// <param name="character"></param>
+		/// <exception cref="InvalidOperationException"></exception>
+		private void SaveCollections(Character character)
+		{
+			using (var conn = this.GetConnection())
+			using (var trans = conn.BeginTransaction())
+			{
+				using (var cmd = new MySqlCommand("DELETE FROM `collections` WHERE `accountId` = @accountId", conn, trans))
+				{
+					cmd.AddParameter("@accountId", character.AccountId);
+					cmd.ExecuteNonQuery();
+				}
+
+				foreach (var collectionId in character.Collections.GetList())
+				{
+					var collectionProgress = character.Collections.GetProgress(collectionId);
+
+					if (collectionProgress.Count > 0)
+					{
+						foreach (var collectionItem in collectionProgress)
+						{
+							using (var cmd = new InsertCommand("INSERT INTO `collections` {0}", conn, trans))
+							{
+								cmd.Set("accountId", character.AccountId);
+								cmd.Set("collectionId", collectionId);
+								cmd.Set("itemId", collectionItem);
+
+								cmd.Execute();
+							}
+						}
+					}
+					else
+					{
+						// we insert a collection with a blank item if you have the collection
```

Comment:
Why is this necessary? Can't the collection just be empty? Especially seeing how you just ignore the -1 entries in loading anyway.

Reply:
Oh, the way it works is that you have to find a hidden treasure chest on each map which gives you an item that unlocks the collection.  The item gets used up, so you'd want the game to remember if you have the collection unlocked, even if it's empty.

Using a separate table is absolutely possible, though I don't think it helps.  The game forces us to send every item in a collection even if it's finished (this is how the game identifies a completed collection), so we still have to query every row in the collection contents table even if we have a separate collections table.  For efficiency, we would probably do this with an outer join (selecting one collection at a time in a loop leads to hundreds of select statements being executed for a single character), but this will now return null for the itemid on an empty collection, forcing us to check for dbNull on every record again.  Overall I think the current implementation is the most optimal.

Reply:
I'll have to take another look at this, but not using a dedicated collections table that connects to a list of items per collection seems like a recipe for refactoring, because the moment we need to save _any_ additional information about a collection, such as a date when it was unlocked for example, we have to redo everything. You're presumably right that it won't give us any boosts in performance, or make the code cleaner in some significant way, but maintainability should always be a concern.

You might've also noticed that the dummy entry implementation is not ideal for conveying intent. Someone who isn't familiar with how exactly the system works will have to dig around the code and database and make several assumptions to come to the conclusion that the dummy actually is a dummy, that's only used to remember a collection is open.